Decouple CES work queue and k8s client rate limiting

dlapcevic opened the pull request:
Decouple CES work queue and k8s client rate limiting

Currently CiliumEndpointSlice (CES) work queue uses rate limiting values specified for the k8s client.

Changes:

```release-note
1. Add a new set of flags for CES work queue limit and burst rates,  `CESWriteQPSLimit` to ` and `CESWriteQPSBurst`.

  The processed work queue items always trigger a single CES create, update or write request to the kube-apiserver.
  The work queue rate limiting effectively limits the rate of writes to the kube-apiserver for CES api objects.

2. Set the default `CESWriteQPSLimit` to `10` and `CESWriteQPSBurst` to `20`.

3. Set the maximums for qps `50` and burst `100`. These values cannot be exceeded regardless of any configuration.

4. Unhide `CESMaxCEPsInCES` and `CESSlicingMode` flags from appearing in logs when `CES` is enabled.
```

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

pchaigno commented on operator/cmd/flags.go:
```diff
@@ -290,13 +290,17 @@ func init() {
 	option.BindEnv(Vp, operatorOption.CESMaxCEPsInCES)
 
 	flags.String(operatorOption.CESSlicingMode, operatorOption.CESSlicingModeDefault, "Slicing mode define how ceps are grouped into a CES")
-	flags.MarkHidden(operatorOption.CESSlicingMode)
```
Do you remember why those were marked hidden? For flags that are present in both the agent and the operator, we often make the operator ones hidden because we don't expect users to use them directly. It will automatically get passed to the operator if they rely on the ConfigMap.

For that reason, I'd expect `EnableCiliumEndpointSlice` above to also be marked hidden.

dlapcevic replied:
I assume they were hidden because they aren’t used by default -- `EnableCiliumEndpointSlice` is false by default.
I see that might be a good reason to hide them.

However, I find it useful to see these flags on operator startup.
It can help with debugging and troubleshooting to see all flags and their values on startup.

Do you think it makes sense to get the value of `EnableCiliumEndpointSlice` and to hide the following 4 flags only if it's `false`?

pchaigno replied:
> I assume they were hidden because they aren’t used by default -- `EnableCiliumEndpointSlice` is false by default.
> I see that might be a good reason to hide them.

We don't usually mark flags hidden because they are unused by default.

> It can help with debugging and troubleshooting to see all flags and their values on startup.

Aren't all those flags already present in agent logs?

> Do you think it makes sense to get the value of EnableCiliumEndpointSlice and to hide the following 4 flags only if it's false?

I think it would be worth bringing that discussion on #development to get the opinion of others. Seems fine to me, although we might want to do that for other features as well.

I don't think this is blocking for this PR in any case.

dlapcevic replied:
Only the `EnableCiliumEndpointSlice` flag is present in agent logs, but the other 4 flags aren't, because they are specific to operators' logic on how to batch CiliumEndpoints into slices.
Agent doesn't care about it.

Even if `EnableCiliumEndpointSlice` flag exists for both operator and agent, there are cases when we wouldn't want the flags to match.
For example, when we are migrating to CEP batching (to use slices), we might want to enable the flag in the operator first, so it generates all the slices, before agents start watching them (still uses CEPs). And only activate it later for agents.

Thank you for the review. I will bring this up in the development channel on slack.

pchaigno replied:
> Only the `EnableCiliumEndpointSlice` flag is present in agent logs, but the other 4 flags aren't, because they are specific to operators' logic on how to batch CiliumEndpoints into slices.

Ah, ok. Then yeah, they probably shouldn't be hidden.

Weil0ng replied:
Just to provide some context I think these ones were marked hidden in the first place because users who are not familiar with how CES works can hardly tune these values to their benefits and a bad value in these arguments could easily break the cluster... but as we add more tests and CES is becoming more battled, I'm not opposed to open these :)

pchaigno reviewed (approved):
One minor comment below. Other than that, LGTM.

nebril reviewed (approved):
LGTM

Weil0ng commented on operator/cmd/flags.go:
```diff
@@ -287,13 +287,17 @@ func init() {
-	flags.MarkHidden(operatorOption.CESSlicingMode)
 	option.BindEnv(Vp, operatorOption.CESSlicingMode)
 
+	flags.Float64(operatorOption.CESWriteQPSLimit, operatorOption.CESWriteQPSLimitDefault, "CES work queue rate limit")
```
I think in the code we should set a limit on this, same as below, otherwise an arbitrary number form user might break their cluster easily.

dlapcevic replied:
Thank you @Weil0ng for the review and suggestion.

You are right. Most users wouldn’t know how to set CES flags to their benefits.
I think it’s anyway good to show the flags when `EnableCiliumEndpointSlice` is turned on, as it might benefit users who are willing to inquire and learn about it.
Do you think it’s a good idea to show the related flags only when CES is enabled?

Users could have misconfigured CES processing/write rate anyway with `K8sClientQPS` in operator.
And I think it isn't easy to break a cluster by misconfiguring it, because the cluster needs to run at scale in the first place (many nodes, high pod churn rate, many pods in total), and that’s rare.

I anyway agree that we should add a hard limit, but I’m not sure which number it should be.
We observed that it could be 100+ QPS for smaller clusters, but 10 QPS for very large clusters, so even with a hard limit it’s possible that users could break their clusters.
Do you have a suggestion?

dlapcevic replied:
I added the maximums for qps `50` and burst `100` in case it's specified more.
I made the change so the CES related flags are only shown when CES is enabled.

Weil0ng replied:
Thank @dlapcevic ! I agree that it is hard (if possible) to pick a limit that suits all cases, maybe a better approach is to have a step function that limit the QPS based on number of nodes in the cluster, but I don't think it should block this PR anyways :)

Weil0ng replied:
nit: since we added a hard limit, maybe it's worthing mentioning in the description text for the flag that it cannot exceed X?

dlapcevic replied:
Done.

dlapcevic committed:
Decouple CES work queue and k8s client rate limiting

Currently CiliumEndpointSlice (CES) work queue uses rate limiting values specified for the k8s client.

Changes:

1. Add a new set of flags for CES work queue limit and burst rates,  `CESWriteQPSLimit` to ` and `CESWriteQPSBurst`.

  The processed work queue items always trigger a single CES create, update or write request to the kube-apiserver.
  The work queue rate limiting effectively limits the rate of writes to the kube-apiserver for CES api objects.

2. Set the default `CESWriteQPSLimit` to `10` and `CESWriteQPSBurst` to `20`.

3. Set the maximums for qps `50` and burst `100`.

4. Unhide `CESMaxCEPsInCES` and `CESSlicingMode` flags from appearing in logs when `CES` is enabled.

Signed-off-by: Dorde Lapcevic <dordel@google.com>
